### ISSUE TITLE
Something problem with batch QC

### ISSUE THREAD

**Issue by jiangli2941** [STOmics/Stereopy#360]
I imported six file merge analysis, their format is in order as follows:
ms_data: {'0': (31381, 23559), '1': (24041, 23059), '2': (26034, 23518), '3': (28167, 23087), '4': (29321, 23074), '5': (29299, 23810)}
num_slice: 6
names: ['0', '1', '2', '3', '4', '5']

I followed the tutorial step by step, but there was a long analysis and pause at this step
ms_data.tl.batch_qc(scope=slice_generator[:],mode='integrate', cluster_res_key='leiden', report_path='./batch_qc', res_key='batch_qc')
Output:
[2024-12-05 22:47:24][Stereo][3971300][MainThread][131909463319616][ms_pipeline][113][INFO]: register algorithm batch_qc to <class 'stereo.core.stereo_exp_data.StereoExpData'>-131907745263232
[2024-12-05 23:03:08][Stereo][3971300][MainThread][131909463319616][classifier][144][INFO]: Model Training Finished!
[2024-12-05 23:03:08][Stereo][3971300][MainThread][131909463319616][classifier][145][INFO]: Trained checkpoint file has been saved to ./batch_qc

Due to the analysis on the cloud server, I waited for a night and did not get the corresponding output result, please ask: 1. Is there a code to simplify the output? 2. Is there a way to improve the speed of operation?

**Comment by jiangli2941** [comment]
Another key question I want to ask is, how do you manually annotate subpopulations of cells? The tutorial only explains singleR's automatic annotation, which is mainly capable of distinguishing immune cells of PBMC. How to manually annotate other parenchymal cells, such as kidney CD-PC,PODO,EC, etc., through marker clustering?

**Comment by tanliwei-coder** [comment]
> Another key question I want to ask is, how do you manually annotate subpopulations of cells? The tutorial only explains singleR's automatic annotation, which is mainly capable of distinguishing immune cells of PBMC. How to manually annotate other parenchymal cells, such as kidney CD-PC,PODO,EC, etc., through marker clustering?

I think if you have a reference of parenchymal cells, singleR also could be used to annotate automatically.

**Comment by tanliwei-coder** [comment]
> I imported six file merge analysis, their format is in order as follows: ms_data: {'0': (31381, 23559), '1': (24041, 23059), '2': (26034, 23518), '3': (28167, 23087), '4': (29321, 23074), '5': (29299, 23810)} num_slice: 6 names: ['0', '1', '2', '3', '4', '5']
> 
> I followed the tutorial step by step, but there was a long analysis and pause at this step ms_data.tl.batch_qc(scope=slice_generator[:],mode='integrate', cluster_res_key='leiden', report_path='./batch_qc', res_key='batch_qc') Output: [2024-12-05 22:47:24][Stereo][3971300][MainThread][131909463319616][ms_pipeline][113][INFO]: register algorithm batch_qc to <class 'stereo.core.stereo_exp_data.StereoExpData'>-131907745263232 [2024-12-05 23:03:08][Stereo][3971300][MainThread][131909463319616][classifier][144][INFO]: Model Training Finished! [2024-12-05 23:03:08][Stereo][3971300][MainThread][131909463319616][classifier][145][INFO]: Trained checkpoint file has been saved to ./batch_qc
> 
> Due to the analysis on the cloud server, I waited for a night and did not get the corresponding output result, please ask: 1. Is there a code to simplify the output? 2. Is there a way to improve the speed of operation?

In the same directory of the notebook you ran the BatchQC, there is a subdirectory called **batch_qc**, in which there is an html file called **BatchQC_reprot_raw.html**, it can be opened directly on notebook.

**Comment by jiangli2941** [comment]
> > I imported six file merge analysis, their format is in order as follows: ms_data: {'0': (31381, 23559), '1': (24041, 23059), '2': (26034, 23518), '3': (28167, 23087), '4': (29321, 23074), '5': (29299, 23810)} num_slice: 6 names: ['0', '1', '2', '3', '4', '5']
> > I followed the tutorial step by step, but there was a long analysis and pause at this step ms_data.tl.batch_qc(scope=slice_generator[:],mode='integrate', cluster_res_key='leiden', report_path='./batch_qc', res_key='batch_qc') Output: [2024-12-05 22:47:24][Stereo][3971300][MainThread][131909463319616][ms_pipeline][113][INFO]: register algorithm batch_qc to <class 'stereo.core.stereo_exp_data.StereoExpData'>-131907745263232 [2024-12-05 23:03:08][Stereo][3971300][MainThread][131909463319616][classifier][144][INFO]: Model Training Finished! [2024-12-05 23:03:08][Stereo][3971300][MainThread][131909463319616][classifier][145][INFO]: Trained checkpoint file has been saved to ./batch_qc
> > Due to the analysis on the cloud server, I waited for a night and did not get the corresponding output result, please ask: 1. Is there a code to simplify the output? 2. Is there a way to improve the speed of operation?
> 
> In the same directory of the notebook you ran the BatchQC, there is a subdirectory called **batch_qc**, in which there is an html file called **BatchQC_reprot_raw.html**, it can be opened directly on notebook.

Thank u for the guiding! Could I ask in deep about the detail in which we can establish the private singleR reference? In R with Seurat package, I usually take use of some cell marker and draw the dotplot, how about in Stereopy? Look forward to reply

**Comment by jiangli2941** [comment]
I eventually tried the singleR annovation with MouseRNAseqData. As I expected, it did perform very badly in the annotation of kidney cells. I tried to convert the data to rda format and annotate the cells in R studio, using the scCATCH package (a annotation toolkit based on single cell clusters, from cluster marker gene identification to cluster annotation based on evidence scoring). This could show some reasonably good comment results.

So the question is, how can you build an individual's singR annotated gene set for matching?

I tried 3 methods:

1. Export the cell_table of the scCATCH package and create it as h5ad file, but it fails because the remaining necessary information of singleR is missing

2. MouseRNAseqData in singR was replaced with gene and cell information in cell_table, but the replacement failed because the number of rows did not match.

3. Export the stereopy standard file to h5ad and use scanpy for annotation, but the export file seems to lack some necessary information, and the annotation still fails.

In short, I really hope to get the author's help in the annotation, I think this is part of the distress after choosing your company's service.

**Comment by tanliwei-coder** [comment]
> > > Another key question I want to ask is, how do you manually annotate subpopulations of cells? The tutorial only explains singleR's automatic annotation, which is mainly capable of distinguishing immune cells of PBMC. How to manually annotate other parenchymal cells, such as kidney CD-PC,PODO,EC, etc., through marker clustering?
> > 
> > 
> > I think if you have a reference about parenchymal cells, singleR also could be used to annotate automatically.
> 
> > > I imported six file merge analysis, their format is in order as follows: ms_data: {'0': (31381, 23559), '1': (24041, 23059), '2': (26034, 23518), '3': (28167, 23087), '4': (29321, 23074), '5': (29299, 23810)} num_slice: 6 names: ['0', '1', '2', '3', '4', '5']
> > > I followed the tutorial step by step, but there was a long analysis and pause at this step ms_data.tl.batch_qc(scope=slice_generator[:],mode='integrate', cluster_res_key='leiden', report_path='./batch_qc', res_key='batch_qc') Output: [2024-12-05 22:47:24][Stereo][3971300][MainThread][131909463319616][ms_pipeline][113][INFO]: register algorithm batch_qc to <class 'stereo.core.stereo_exp_data.StereoExpData'>-131907745263232 [2024-12-05 23:03:08][Stereo][3971300][MainThread][131909463319616][classifier][144][INFO]: Model Training Finished! [2024-12-05 23:03:08][Stereo][3971300][MainThread][131909463319616][classifier][145][INFO]: Trained checkpoint file has been saved to ./batch_qc
> > > Due to the analysis on the cloud server, I waited for a night and did not get the corresponding output result, please ask: 1. Is there a code to simplify the output? 2. Is there a way to improve the speed of operation?
> > 
> > 
> > In the same directory of the notebook you ran the BatchQC, there is a subdirectory called **batch_qc**, in which there is an html file called **BatchQC_reprot_raw.html**, it can be opened directly on notebook.
> 
> But I only found to look at the suffix bgi batchQC file, does this mean that the output failed? ![2400698f-bf53-4fe5-93fe-a7bde8597e42](https://private-user-images.githubusercontent.com/44805254/393879975-b8e81f34-f318-4c17-b0f9-5500309964f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUyODE5MTgsIm5iZiI6MTczNTI4MTYxOCwicGF0aCI6Ii80NDgwNTI1NC8zOTM4Nzk5NzUtYjhlODFmMzQtZjMxOC00YzE3LWIwZjktNTUwMDMwOTk2NGY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI3VDA2NDAxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkMTA5NTU0YTc0NzNhMmIzM2VhNjY0ZWU4MjQ5ZWEwZGUwNzM4MmFjMjI2Njg1YjRmMjU2YzUwNzhmNmIxMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WaQUHZ0_oDuWjG0WHLfPDp4k_QjfBUNlZC5uoa41TiA)

I think your data is so lager that the BatchQC tend to take more time, I don't have your data, I can not judge it correctly.

**Comment by tanliwei-coder** [comment]
The reference only needs to be an h5ad file in which the obs has a column of cell annotation, h5ad is a file format used to save AnnData, when you run the singleR, set the parameter `ref_use_col` to the obs column name of the cell annotation

**Comment by tanliwei-coder** [comment]
There is also a method, you can use the [clustering methods](https://stereopy.readthedocs.io/en/latest/Tutorials/CellBin_Clustering.html) of stereopy to cluster the data you want to annotate, and use method [data.tl.annotation](https://stereopy.readthedocs.io/en/latest/content/stereo.core.StPipeline.annotation.html#stereo.core.StPipeline.annotation) to annotate the cluster result manually.
